### PR TITLE
Ecforce ECS Credentials Provider IP Rules

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- Adds logic to enforce the ECS provider ip address rules. https://github.com/awslabs/aws-c-io/pull/639 adds test which validates that the host resolver will return the shortest form IPv6 which makes it easier to assert equality for IPv6 in this PR. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/credentials_provider_ecs.c [hunk 1]
```diff
@@ -459,6 +464,90 @@ static void s_ecs_on_acquire_connection(struct aws_http_connection *connection,
     s_ecs_query_task_role_credentials(ecs_user_data);
 }
 
+/*
+ * The resolved IP address must satisfy one of the following:
+ * 1. within the loopback CIDR (IPv4 127.0.0.0/8, IPv6 ::1/128)
+ * 2. corresponds to the ECS container host 169.254.170.2
+ * 3. corresponds to the EKS container host IPs (IPv4 169.254.170.23, IPv6 fd00:ec2::23)
+ */
+static bool s_is_valid_remote_host_ip(struct aws_host_address *host_address_ptr) {
```
Comment (graebm) [inline source/credentials_provider_ecs.c:473]: utterly trivial: 
```suggestion
static bool s_is_valid_remote_host_ip(const struct aws_host_address *host_address) {
```

File: source/credentials_provider_ecs.c [hunk 2]
```diff
@@ -459,6 +464,90 @@ static void s_ecs_on_acquire_connection(struct aws_http_connection *connection,
...
+        const struct aws_byte_cursor ecs_container_host_address = aws_byte_cursor_from_c_str("169.254.170.2");
+        const struct aws_byte_cursor eks_container_host_address = aws_byte_cursor_from_c_str("169.254.170.23");
+
+        result |= aws_byte_cursor_starts_with(&address, &ipv4_loopback_address_prefix);
+        result |= aws_byte_cursor_eq(&address, &ecs_container_host_address);
+        result |= aws_byte_cursor_eq(&address, &eks_container_host_address);
+
+    } else if (host_address_ptr->record_type == AWS_ADDRESS_RECORD_TYPE_AAAA) { /* Check for both the short form and
+                                                                                long
+                                                                                form of an IPv6 address to be safe. */
```
Comment (graebm) [inline source/credentials_provider_ecs.c:487]: utterly trivial
```suggestion
    } else if (host_address_ptr->record_type == AWS_ADDRESS_RECORD_TYPE_AAAA) {
        /* Check for both the short form and long form of an IPv6 address to be safe. */
```

**Review by graebm — APPROVED** [review]
just nitpicks. looks good!

File: include/aws/auth/credentials.h [hunk 3]
```diff
@@ -232,6 +232,15 @@ struct aws_credentials_provider_imds_options {
  * AWS_CONTAINER_CREDENTIALS_RELATIVE_URI
  * AWS_CONTAINER_CREDENTIALS_FULL_URI
  *
+ *`AWS_CONTAINER_CREDENTIALS_FULL_URI` URL must satisfy one of the following:
+ *  1. The URL begins with `https`.
+ *  2. The URL refers to an allowed IP address. If a URL contains a domain name instead of an IP address,
+ *      a DNS lookup will be performed. All resolved IP addresses must refer to an allowed IP address, or
+ *      the credentials provider will return `AWS_AUTH_CREDENTIALS_PROVIDER_ECS_INVALID_HOST`. Valid IP addresses are:
```
[the remark this answers is not in the document]
Comment (waahm7) [reply include/aws/auth/credentials.h]: Thanks, I have updated it to get rid of the nested rules list. Hopefully, it's clearer now.

**Pull request merged**